src/components/tl/cuda/reduce_scatterv/reduce_scatterv_linear.c

```diff
@@ -0,0 +1,443 @@
+    int                     num_steps = nfrags + 1;
+    ucc_ee_executor_task_t      *etask;
+    ucc_status_t                 st;
+    ucc_ee_executor_t           *exec;
+    int                          step, i;
+    void                        *sbuf, *rbuf;
+    size_t local_offset, remote_offset, rank_offset;
+
+    st = ucc_coll_task_get_executor(&task->super, &exec);
+    if (ucc_unlikely(st != UCC_OK)) {
+        return st;
+    }
+
+    step = get_rank_step(task, trank, 0);
+    while (step < num_steps) {
+        if ((task->reduce_scatterv_linear.exec_task[0] != NULL) ||
+            (task->reduce_scatterv_linear.exec_task[1] != NULL)) {
+            for (i = 0; i < 2; i++) {
+                etask = task->reduce_scatterv_linear.exec_task[i];
+                if (etask != NULL) {
+                    st = ucc_ee_executor_task_test(etask);
+                    if (st == UCC_OK) {
+                        ucc_ee_executor_task_finalize(etask);
+                        task->reduce_scatterv_linear.exec_task[i] = NULL;
+                    } else {
+                        if (ucc_likely(st > 0)) {
+                            return UCC_INPROGRESS;
+                        }
+                        return st;
+                    }
+                }
+            }
+            step++;
+            set_rank_step(task, trank, step, 0);
+            continue;
+        }
+
+        for (i = 0; i < tsize; i++) {
+            if (get_rank_step(task, i, 0) < step) {
+                return UCC_INPROGRESS;
+            }
+        }
+
+        if (step % 2) {
+            remote_offset = ssize / 2;
+            local_offset  = 0;
+        } else {
+            remote_offset = 0;
+            local_offset  = ssize / 2;
+        }
+
+        if (UCC_IS_INPLACE(*args)) {
+            rank_offset = task->reduce_scatterv_linear.get_offset(task, trank);
+            sbuf = task->reduce_scatterv_linear.rbuf;
+            rbuf = PTR_OFFSET(sbuf, rank_offset * ucc_dt_size(dt));
+        } else {
+            sbuf = task->reduce_scatterv_linear.sbuf;
+            rbuf = task->reduce_scatterv_linear.rbuf;
+        }
+
+        if (step == 0) {
+            ucc_tl_cuda_reduce_scatterv_linear_copy(task, exec, sbuf, step,
+                remote_offset, &task->reduce_scatterv_linear.exec_task[0]);
+        } else if (step == (num_steps - 1)) {
+            ucc_tl_cuda_reduce_scatterv_linear_reduce(task, exec, sbuf, rbuf,
+                step - 1, local_offset, &task->reduce_scatterv_linear.exec_task[1]);
+        } else {
+            ucc_tl_cuda_reduce_scatterv_linear_copy(task, exec, sbuf, step,
+                remote_offset, &task->reduce_scatterv_linear.exec_task[0]);
+            ucc_tl_cuda_reduce_scatterv_linear_reduce(task, exec, sbuf, rbuf,
+                step - 1, local_offset, &task->reduce_scatterv_linear.exec_task[1]);
+        }
+
+        if ((task->reduce_scatterv_linear.exec_task[0] == NULL) &&
```

Comment:
If we remove the if condition on line 259-260 I think we can get rid of this block

Reply:
can you pls elaborate what condition to use in lines 259-260 instead?

Reply:
I suggest to completely remove the if condition (and use no condition at all). If both `task[0]` and `task[1]` are `NULL`, then we don't progress any worker and just do
```
step++;
set_rank_step(task, trank, step, 0);
```
so we don't need to handle this case as a special case as it is done currently
Am I correct ?